### PR TITLE
Remodel the network layer

### PR DESCRIPTION
**Pull request by edsko**
Most of this stuff I already discussed with @brunjlar (protocol changes) and with @dcoutts (network abstraction changes).

### CODE REVIEW

File: io-sim-classes/src/Control/Monad/Class/MonadFork.hs [hunk 1]
```diff
@@ -8,16 +8,25 @@ import           Control.Exception
...
   fork a =
     let handleException :: Either SomeException () -> IO ()
-        handleException (Left e) =
-            hPutStrLn stderr $ "Uncaught exception in thread:" ++ displayException e
+        handleException (Left e) = do
+            tid <- IO.myThreadId
+            IO.withMVar forkPrintExceptionLock $ \() -> do
+              hPutStrLn stderr $ "Uncaught exception in thread " ++ show tid
+                              ++ ": " ++ displayException e
+              hFlush stderr
```
Comment (avieth) [inline io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: This isn't a criticism of the change here, but of having a default exception handler in general. With or without this change, the user has no way to stop this print to `stderr`. If you want exception logging, log the exception yourself, it should not be the default built-in unalterable definition of a `fork`.

File: io-sim-classes/src/Control/Monad/Class/MonadSTM.hs [hunk 2]
```diff
@@ -44,7 +46,7 @@ class (MonadFork m, Monad (Tr m)) => MonadSTM m where
   -- The STM primitives
   type TVar m :: * -> *
 
-  atomically   :: Tr m a -> m a
+  atomically   :: HasCallStack => Tr m a -> m a
```
Comment (avieth) [inline io-sim-classes/src/Control/Monad/Class/MonadSTM.hs]: Will this mean that every user of `atomically` will have to state that constraint?

Why not leave it as is and just export `wrapBlockedIndefinitely`?

File: io-sim-classes/src/Control/Monad/Class/MonadFork.hs (hunk 1 above)
Comment (edsko) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: The point is that if you forget to add an exception handler, the default is to do nothing with it, and you will not notice. I agree that what is here is crude, but I personally find it invalueable when unexpected exceptions are noticed. Obviously printing them is crude, if we had better logging that would be better.

File: io-sim-classes/src/Control/Monad/Class/MonadSTM.hs (hunk 2 above)
Comment (edsko) [reply io-sim-classes/src/Control/Monad/Class/MonadSTM.hs]: No, users of `atomically` will not be affected. They can state `HasCallStack` if they want to grow the recorded call stack but they don't have to. By having it here it means that no matter where `atomically` is called, _if_ we get that "blocked indefinitely" exception we are told where it came from without having to do anything else. Again, very useful for debugging; "blocked indefinitely" exceptions can be hard to track down; now, without doing anything else, you get a line number to go with it.

**Comment by edsko** [comment]
@avieth Since you are objecting less to the _change_ to fork than to having a default exception handler at all, I'm going to merge this now; if you really feel that exceptions should be swallowed by default without trace we can always change it later.

**Pull request merged**